### ISSUE TITLE
adding DGL regression on a graph level

### ISSUE THREAD

**Issue by Sinasta** [wassimj/topologicpy#13]
I was thinking that it could be interesting to add the possibility of training a regression model for predicting graph level values.
Would you be interested in that?
I know that DGL already includes regression models for node/edge prediction but i think it could be interesting to implement that on a graph level basis as mentioned [here](https://discuss.dgl.ai/t/can-dgl-be-used-for-regression-problems-too/2667).
If you think this would be interesting i would be happy to experiment in that direction.

**Comment by wassimj** [comment]
Sure! Go for it!

**Comment by Sinasta** [comment]
Perfect! I will have a look into that probably today. I will reopen this issue if i encounter any problems.

**Comment by Sinasta** [comment]
I have pushed a preliminary implementation of graph level regression to my [fork](https://github.com/Sinasta/topologicpy).
It is somehow working (only for cv_type: holdout for now) and is extremely messy. Basically i had to introduce a variable: `task` which holds a string that can be either 'regression' or 'classification'.
I ended up with a lot of difficulties adapting the rest of your DGL API to those two different ML tasks.
But basically the main changes to switch from classification to regression are:

- changing the type of input (no longer classes but now values) 
- the loss function from cross entropy/negative log likelihood to `F.mse_loss`
- the output dimension of the model to 1
- the final layer of the model to `nn.Linear(dim[-1], 1)`

I sadly don't think that i will have enough free time to finish the implementation in the coming month because i have to write my master thesis but the current implementation is somehow working (far from being ready to be merged though)

The current major problems are:

- ~~the predictions seem to be pretty bad as soon as the batch size is more then 1~~ (seems to be fixed now)
- ~~the regression implementation breaks all the accuracy reports that you built~~
- ~~general respect of the topologicAPI needs to be reworked~~
- K-Fold is not implemented at all
- regression models can maybe be improved

I am writing this in case you are interested to have a look and maybe continue where i had to stop. I am more then happy to answer any question as far as my knowledge allows

**Comment by Sinasta** [comment]
Now that i think about it...it might be better to implement seperate functions for regression (models and train function) in the DGL.py for the sake of code clarity and testing/mantainance.

**Comment by wassimj** [comment]
I agree. I did the same for node classification. Makes it easier to manage. I will take a look soon, but I need to educate myself on how your code works. It will take time as I am juggling a few things at this point.

**Comment by wassimj** [comment]
Can you please send me a sample dataset that I can use for testing? Thanks

**Comment by Sinasta** [comment]
I send you an email with my dataset and some small explanations.

**Comment by wassimj** [comment]
@Sinasta  Did you address the one-hot-encoding that I do for classification of graph labels? How does that code work with your graphs if the label has a continuous value?  What methods do you use to load graphs? Is it just from CSV files, or do you also load them from topologic graphs with embedded dictionaries? (ByGraph, and ByImportedCSV). Thanks

**Comment by wassimj** [comment]
I have pushed a draft and untested version of DGL and a new version of topologicpy 0.3.0. I made some changes to your code:
1. "testing" is now called "validation", and "test" is now called "validate"
2. Classifier and Regressor methods now accept training, validation, and testing datasets (or if not present they use the split [0.8,0.1,0.1] hyper-parameter
3. The training methods are now called "TrainClassifier", "TrainNodeClassifier", and "TrainRegressor"
4. The Classification/Prediction methods are now called "Classify", "ClassifyNode", and "Predict"
5. The Accuracy methods are now called "Accuracy", and "RMSE"
6. The classifier and regressor classes are now called "_ClassifierHoldout", "_ClassifierKFold" and "_RegressorHoldout" The underscore "_" at the start of the name is to hide it form the auto-documentation software.
7. The convolutional models for regression are now called "_ClassicReg" and "_GraphConvReg" I have removed SAGEConv, TAGConv etc as it seems they are not yet compatible with your code.

**Comment by Sinasta** [comment]
> @Sinasta  Did you address the one-hot-encoding that I do for classification of graph labels? How does that code work with your graphs if the label has a continuous value?  What methods do you use to load graphs? Is it just from CSV files, or do you also load them from topologic graphs with embedded dictionaries? (ByGraph, and ByImportedCSV). Thanks

I do not hot encode the graph values because as you said they are continuous values so there is no need for encoding. The node labels are still hot encoded through your implemented function.
For loading of the .bin datasets i use the native DGL graph loading functions.`dgl.load_graphs()`

**Comment by Sinasta** [comment]
I will try out your new changes/regressor inclusion as soon as i find some time to do that. Just a small side note: SageConv, TagConv, GinConv models can be used for regression in my understanding. In my code i did not add extra models for them because there is no need for extra model setups because it is enough to call them with 1 as output layer dimension. If this was not clear i will explain more when i find time.

**Comment by wassimj** [comment]
> SageConv, TagConv, GinConv models can be used for regression in my understanding.

Oh, I see! ok, I will put them back as options then.

**Comment by Sinasta** [comment]
I proposed a PR with some refinements for the Regressor training and a simplification for the Classifier accuracy calculation during the training. To me it seems that the only thing missing for a functional regressor inclusion is to implement K-fold. What's your opinion on that? Did you manage to test the regression with my dataset?

**Comment by wassimj** [comment]
I am revamping the whole class to integrate classifier and regressor. All the train, validate, and test methods are now integrated and you can call them with any model (I now call it model instead of classifier or regressor). Please stay tuned. I think it will be easy to add a K-Fold to the regressor model.

**Comment by wassimj** [comment]
Thank you for contributing code to topologicpy! I think this is solved for now so I will close it. But feel free to re-open if anything is still outstanding.
